terraform_module_pinned_source rule emits unclear error for un-parseable URL

We have an project in AzureDevops which was created with a space in the name, "My Project". When referencing repos within that project, tflint fails on the pinned_source rule, seemingly misreading the end of line.

For this bit of terraform,

```hcl
module "rg" {
  source   = "git::https://myorg:#{ terraformAzureDevOpsPAT }#@dev.azure.com/myorg/My%20Project/_git/terraform-az-rg?ref=0.1.0"
  name     = "rg-${local.application}-${var.environment}-${var.fpbuild_id}"
  location = var.location
  tags     = local.default_tags
}
```

We see this error message,

```
2021-08-24 13:01:40 [ERROR]   Found errors in [tflint] linter!
2021-08-24 13:01:40 [ERROR]   Error code: 3. Command output:
------
1 issue(s) found:

Warning: Module source "git::https://myorg:#{ terraformAzureDevOpsPAT }#@dev.azure.com/myorg/My%!P(MISSING)roject/_git/terraform-az-rg?ref=0.1.0" is not pinned (terraform_module_pinned_source)

  on /tmp/lint/main.tf line 2:
   2:   source   = "git::https://myorg:#{ terraformAzureDevOpsPAT }#@dev.azure.com/myorg/My%20Plus/_git/terraform-az-rg?ref=0.1.0"

Reference: https://github.com/terraform-linters/tflint/blob/v0.31.0/docs/rules/terraform_module_pinned_source.md
```

## Version

```console
> tflint -v
TFLint version 0.31.0
> terraform -v
Terraform v0.14.10
```

--- a/rules/terraformrules/terraform_module_pinned_source.go
+++ b/rules/terraformrules/terraform_module_pinned_source.go
@@ -5,6 +5,7 @@ import (
 	"log"
 	"net/url"
 	"path/filepath"
+	"strings"
 
 	"github.com/Masterminds/semver/v3"
 	"github.com/hashicorp/go-getter"
@@ -91,6 +92,27 @@ func (r *TerraformModulePinnedSourceRule) checkModule(runner *tflint.Runner, mod
 	switch u.Scheme {
 	case "git", "hg":
 	default:
+		return nil
+	}
+
+	if u.Opaque != "" {
+		// for git:: or hg:: pseudo-URLs, Opaque is :https, but query will still be parsed
+		query := u.RawQuery
+		u, err = url.Parse(strings.TrimPrefix(u.Opaque, ":"))
+		if err != nil {
+			return err
+		}
+
+		u.RawQuery = query
+	}
+
+	if u.Hostname() == "" {
+		runner.EmitIssue(
+			r,
+			fmt.Sprintf("Module source %q is not a valid URL", module.SourceAddr),
+			module.SourceAddrRange,
+		)
+
 		return nil
 	}
 

--- a/rules/terraformrules/terraform_module_pinned_source_test.go
+++ b/rules/terraformrules/terraform_module_pinned_source_test.go
@@ -69,6 +69,24 @@ module "pinned_git" {
 			Expected: tflint.Issues{},
 		},
 		{
+			Name: "invalid URL",
+			Content: `
+module "invalid" {
+  source = "git://#{}.com"
+}`,
+			Expected: tflint.Issues{
+				{
+					Rule:    NewTerraformModulePinnedSourceRule(),
+					Message: `Module source "git://#{}.com" is not a valid URL`,
+					Range: hcl.Range{
+						Filename: "module.tf",
+						Start:    hcl.Pos{Line: 3, Column: 12},
+						End:      hcl.Pos{Line: 3, Column: 27},
+					},
+				},
+			},
+		},
+		{
 			Name: "git module reference is pinned, but style is semver",
 			Content: `
 module "pinned_git" {

